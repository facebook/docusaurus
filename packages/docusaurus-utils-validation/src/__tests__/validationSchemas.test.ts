/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Joi from '../Joi';

import {
  AdmonitionsSchema,
  RehypePluginsSchema,
  RemarkPluginsSchema,
  PluginIdSchema,
  URISchema,
  PathnameSchema,
  RouteBasePathSchema,
  ContentVisibilitySchema,
  FrontMatterLastUpdateSchema,
} from '../validationSchemas';

function createTestHelpers({
  schema,
  defaultValue,
}: {
  schema: Joi.Schema;
  defaultValue?: unknown;
}) {
  function testOK(value: unknown, options?: {normalizedValue?: unknown}) {
    const expectedValue = options?.normalizedValue ?? value ?? defaultValue;
    expect(Joi.attempt(value, schema)).toEqual(expectedValue);
  }

  function testFail(value: unknown) {
    expect(() => Joi.attempt(value, schema)).toThrowErrorMatchingSnapshot(
      `for value=${JSON.stringify(value)}`,
    );
  }

  return {testOK, testFail};
}

function testMarkdownPluginSchemas(schema: Joi.Schema) {
  const {testOK, testFail} = createTestHelpers({
    schema,
    defaultValue: [],
  });

  testOK(undefined);
  testOK([() => {}]);
  testOK([[() => {}, {attr: 'val'}]]);
  testOK([[() => {}, {attr: 'val'}], () => {}, [() => {}, {attr: 'val'}]]);
  // cSpell:ignore remarkjs
  // official `remarkjs/remark-frontmatter` plugin accepts string options
  testOK([[() => {}, 'string-option']]);
  testOK([[() => {}, true]]);

  testFail(null);
  testFail(false);
  testFail(3);
  testFail([null]);
  testFail([false]);
  testFail([3]);
  testFail([[]]);
  testFail([[() => {}, undefined]]);
}

describe('validation schemas', () => {
  it('pluginIdSchema', () => {
    const {testOK, testFail} = createTestHelpers({
      schema: PluginIdSchema,
      defaultValue: 'default',
    });

    testOK(undefined);
    testOK('docs');
    testOK('default');
    testOK('plugin-id_with-simple-special-chars');
    testOK('doc1');

    testFail('/docs');
    testFail('docs/');
    testFail('do/cs');
    testFail('do cs');
    testFail(null);
    testFail(3);
    testFail(true);
    testFail([]);
  });

  it('admonitionsSchema', () => {
    const {testOK, testFail} = createTestHelpers({
      schema: AdmonitionsSchema,
      defaultValue: true,
    });

    testOK(undefined);
    testOK(true);
    testOK(false);
    testOK({});
    testOK({keywords: ['info', 'tip']});
    testOK({keywords: ['info', 'tip'], extendDefaults: true});
    testOK({keywords: ['info', 'tip'], extendDefaults: false});
    testOK({keywords: []});
    testOK({keywords: [], extendDefaults: true}); // noop
    testOK({keywords: [], extendDefaults: false}); // disable admonitions
    testOK({keywords: ['custom-keyword'], extendDefaults: true});
    testOK({keywords: ['custom-keyword'], extendDefaults: false});

    testFail(3);
    testFail([]);
    testFail({unknownAttribute: 'val'});
    testFail({tag: ''});
    testFail({keywords: ['custom-keyword'], extendDefaults: 42});
    testFail({tag: '+++'});
    testFail({tag: '+++', keywords: ['info', 'tip']});

    // Legacy types
    testFail({
      infima: true,
    });
    testFail({
      icons: 'emoji',
    });
    testFail({
      customTypes: {
        myKeyword: {
          keyword: `myKeyword`,
          infima: true,
          svg: '<svg width="512px" height="512px" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"></svg>',
        },
      },
    });
  });

  it('remarkPluginsSchema', () => {
    testMarkdownPluginSchemas(RemarkPluginsSchema);
  });

  it('rehypePluginsSchema', () => {
    testMarkdownPluginSchemas(RehypePluginsSchema);
  });

  it('uRISchema', () => {
    const {testFail, testOK} = createTestHelpers({schema: URISchema});

    const validURL = 'https://docusaurus.io';
    const doubleHash = 'https://docusaurus.io#github#/:';
    const invalidURL = 'spaces are invalid in a URL';
    const relativeURL = 'relativeURL';
    const relativeURLWithParent = '../relativeURLWithParent';
    const urlFromIssue = 'https://riot.im/app/#/room/#ligo-public:matrix.org';
    testOK(validURL);
    testOK(doubleHash);
    testFail(invalidURL);
    testOK(relativeURL);
    testOK(relativeURLWithParent);
    testOK(urlFromIssue);

    const protocolRelativeUrl1 = '//docusaurus.io/path';
    const protocolRelativeUrl2 = '//docusaurus.io/docs/doc1#hash';
    testOK(protocolRelativeUrl1);
    testOK(protocolRelativeUrl2);
  });

  it('pathnameSchema', () => {
    const {testFail, testOK} = createTestHelpers({schema: PathnameSchema});

    testOK('/foo');
    testFail('foo');
    testFail('https://github.com/foo');
  });

  it('routeBasePathSchema', () => {
    const {testFail, testOK} = createTestHelpers({
      schema: RouteBasePathSchema,
      defaultValue: undefined,
    });

    testOK('', {normalizedValue: '/'});
    testOK('/');
    testOK('/foo', {normalizedValue: '/foo'});
    testOK('foo', {normalizedValue: '/foo'});
    testOK('blog', {normalizedValue: '/blog'});
    testOK('blog/', {normalizedValue: '/blog/'});
    testOK('prefix/blog', {normalizedValue: '/prefix/blog'});
    testOK('prefix/blog/', {normalizedValue: '/prefix/blog/'});
    testOK('/prefix/blog', {normalizedValue: '/prefix/blog'});
    testOK(undefined);

    testFail(3);
    testFail([]);
    testFail(null);
    testFail({});
  });

  it('contentVisibilitySchema', () => {
    const {testFail, testOK} = createTestHelpers({
      schema: ContentVisibilitySchema,
    });

    testOK({});
    testOK({draft: false});
    testOK({draft: true});
    testOK({unlisted: false});
    testOK({unlisted: true});

    testOK({draft: false, unlisted: false});
    testOK({draft: true, unlisted: false});
    testOK({draft: false, unlisted: true});
    testOK({draft: true, unlisted: undefined});
    testOK({draft: undefined, unlisted: true});

    testFail({draft: 'bad string'});
    testFail({draft: 42});
    testFail({unlisted: 'bad string'});
    testFail({unlisted: 42});
    testFail({draft: true, unlisted: true});
  });

  it('frontMatterLastUpdateSchema schema', () => {
    const {testFail, testOK} = createTestHelpers({
      schema: FrontMatterLastUpdateSchema,
    });

    testOK(undefined);
    testOK({date: '2021-01-01'});
    testOK({date: '2021-01'});
    testOK({date: '2021'});
    testOK({date: new Date()});
    testOK({author: 'author'});
    testOK({author: 'author', date: '2021-01-01'});
    testOK({author: 'author', date: new Date()});

    testFail(null);
    testFail({});
    testFail('string');
    testFail(42);
    testFail(true);
    testFail([]);
    testFail({author: 23});
    testFail({date: '20-20-20'});
  });
});
