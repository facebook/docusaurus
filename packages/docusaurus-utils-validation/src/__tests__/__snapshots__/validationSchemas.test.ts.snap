// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation schemas admonitionsSchema: for value=[] 1`] = `""value" does not look like a valid admonitions config"`;

exports[`validation schemas admonitionsSchema: for value={"customTypes":{"myKeyword":{"keyword":"myKeyword","infima":true,"svg":"<svg width=\\"512px\\" height=\\"512px\\" viewBox=\\"0 0 512 512\\" xmlns=\\"http://www.w3.org/2000/svg\\"></svg>"}}} 1`] = `""customTypes" is not allowed"`;

exports[`validation schemas admonitionsSchema: for value={"icons":"emoji"} 1`] = `""icons" is not allowed"`;

exports[`validation schemas admonitionsSchema: for value={"infima":true} 1`] = `""infima" is not allowed"`;

exports[`validation schemas admonitionsSchema: for value={"keywords":["custom-keyword"],"extendDefaults":42} 1`] = `""extendDefaults" must be a boolean"`;

exports[`validation schemas admonitionsSchema: for value={"tag":""} 1`] = `"It is not possible anymore to use a custom admonition tag. The only admonition tag supported is ':::' (Markdown Directive syntax)"`;

exports[`validation schemas admonitionsSchema: for value={"tag":"+++","keywords":["info","tip"]} 1`] = `"It is not possible anymore to use a custom admonition tag. The only admonition tag supported is ':::' (Markdown Directive syntax)"`;

exports[`validation schemas admonitionsSchema: for value={"tag":"+++"} 1`] = `"It is not possible anymore to use a custom admonition tag. The only admonition tag supported is ':::' (Markdown Directive syntax)"`;

exports[`validation schemas admonitionsSchema: for value={"unknownAttribute":"val"} 1`] = `""unknownAttribute" is not allowed"`;

exports[`validation schemas admonitionsSchema: for value=3 1`] = `""value" does not look like a valid admonitions config"`;

exports[`validation schemas contentVisibilitySchema: for value={"draft":"bad string"} 1`] = `""draft" must be a boolean"`;

exports[`validation schemas contentVisibilitySchema: for value={"draft":42} 1`] = `""draft" must be a boolean"`;

exports[`validation schemas contentVisibilitySchema: for value={"draft":true,"unlisted":true} 1`] = `"Can't be draft and unlisted at the same time."`;

exports[`validation schemas contentVisibilitySchema: for value={"unlisted":"bad string"} 1`] = `""unlisted" must be a boolean"`;

exports[`validation schemas contentVisibilitySchema: for value={"unlisted":42} 1`] = `""unlisted" must be a boolean"`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value="string" 1`] = `""value" does not look like a valid last update object. Please use an author key with a string or a date with a string or Date."`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value=[] 1`] = `""value" does not look like a valid last update object. Please use an author key with a string or a date with a string or Date."`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value={"author":23} 1`] = `""author" must be a string"`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value={"date":"20-20-20"} 1`] = `""date" must be a valid date"`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value={} 1`] = `""value" does not look like a valid last update object. Please use an author key with a string or a date with a string or Date."`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value=42 1`] = `""value" does not look like a valid last update object. Please use an author key with a string or a date with a string or Date."`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value=null 1`] = `""value" does not look like a valid last update object. Please use an author key with a string or a date with a string or Date."`;

exports[`validation schemas frontMatterLastUpdateSchema schema: for value=true 1`] = `""value" does not look like a valid last update object. Please use an author key with a string or a date with a string or Date."`;

exports[`validation schemas pathnameSchema: for value="foo" 1`] = `""value" (foo) is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;

exports[`validation schemas pathnameSchema: for value="https://github.com/foo" 1`] = `""value" (https://github.com/foo) is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;

exports[`validation schemas pluginIdSchema: for value="/docs" 1`] = `"Illegal plugin ID value "/docs": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value="do cs" 1`] = `"Illegal plugin ID value "do cs": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value="do/cs" 1`] = `"Illegal plugin ID value "do/cs": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value="docs/" 1`] = `"Illegal plugin ID value "docs/": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value=[] 1`] = `""value" must be a string"`;

exports[`validation schemas pluginIdSchema: for value=3 1`] = `""value" must be a string"`;

exports[`validation schemas pluginIdSchema: for value=null 1`] = `""value" must be a string"`;

exports[`validation schemas pluginIdSchema: for value=true 1`] = `""value" must be a string"`;

exports[`validation schemas rehypePluginsSchema: for value=[[]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[[null,null]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[3] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[false] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[null] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=3 1`] = `""value" must be an array"`;

exports[`validation schemas rehypePluginsSchema: for value=false 1`] = `""value" must be an array"`;

exports[`validation schemas rehypePluginsSchema: for value=null 1`] = `""value" must be an array"`;

exports[`validation schemas remarkPluginsSchema: for value=[[]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[[null,null]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[3] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[false] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[null] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=3 1`] = `""value" must be an array"`;

exports[`validation schemas remarkPluginsSchema: for value=false 1`] = `""value" must be an array"`;

exports[`validation schemas remarkPluginsSchema: for value=null 1`] = `""value" must be an array"`;

exports[`validation schemas routeBasePathSchema: for value=[] 1`] = `""value" must be a string"`;

exports[`validation schemas routeBasePathSchema: for value={} 1`] = `""value" must be a string"`;

exports[`validation schemas routeBasePathSchema: for value=3 1`] = `""value" must be a string"`;

exports[`validation schemas routeBasePathSchema: for value=null 1`] = `""value" must be a string"`;

exports[`validation schemas uRISchema: for value="spaces are invalid in a URL" 1`] = `""value" does not look like a valid url (value='')"`;
